web_src/js/features/user-auth-webauthn.js

```diff
@@ -0,0 +1,243 @@
+        window.location.href = '/';
+      }
+    },
+    error: (ret) => {
+      console.error('3-----', ret);
+    }
+  });
+}
+
+const lookup = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
+
+function encode (num) {
+  return lookup.charAt(num);
+}
+
+function tripletToBase64 (num) {
+  return encode(num >> 18 & 0x3F) + encode(num >> 12 & 0x3F) + encode(num >> 6 & 0x3F) + encode(num & 0x3F);
+}
+
+function uint8ToBase64 (uint8) {
+  const extraBytes = uint8.length % 3; // if we have 1 byte left, pad 2 bytes
+  let output = '';
+  let temp = 0;
+
+  // go through the array every three bytes, we'll deal with trailing stuff later
+  for (let i = 0, length = uint8.length - extraBytes; i < length; i += 3) {
+    temp = (uint8[i] << 16) + (uint8[i + 1] << 8) + (uint8[i + 2]);
+    output += tripletToBase64(temp);
+  }
+
+  // pad the end with zeros, but make sure to not forget the extra bytes
+  switch (extraBytes) {
+    case 1:
+      temp = uint8[uint8.length - 1];
+      output += encode(temp >> 2);
+      output += encode((temp << 4) & 0x3F);
+      output += '==';
+      break;
+    case 2:
+      temp = (uint8[uint8.length - 2] << 8) + (uint8[uint8.length - 1]);
+      output += encode(temp >> 10);
+      output += encode((temp >> 4) & 0x3F);
+      output += encode((temp << 2) & 0x3F);
+      output += '=';
+      break;
+    default:
+      break;
+  }
+
+  return output;
+}
+
+// Encode an ArrayBuffer into a base64 string.
+function bufferEncode(value) {
+  return uint8ToBase64(value)
+    .replace(/\+/g, '-')
+    .replace(/\//g, '_')
+    .replace(/=/g, '');
+}
+
+function webauthnRegistered(newCredential) {
+  if (checkError(newCredential)) {
+    return;
+  }
+
+  const attestationObject = new Uint8Array(newCredential.response.attestationObject);
+  const clientDataJSON = new Uint8Array(newCredential.response.clientDataJSON);
+  const rawId = new Uint8Array(newCredential.rawId);
+
+  $.ajax({
+    url: `${appSubUrl}/user/settings/security/webauthn/register`,
+    type: 'POST',
+    headers: {'X-Csrf-Token': csrfToken},
+    data: JSON.stringify({
+      id: newCredential.id,
+      rawId: bufferEncode(rawId),
+      type: newCredential.type,
+      response: {
+        attestationObject: bufferEncode(attestationObject),
+        clientDataJSON: bufferEncode(clientDataJSON),
+      },
+    }),
+    dataType: 'json',
+    contentType: 'application/json; charset=utf-8',
+    success() {
+      window.location.reload();
+    },
+    fail() {
+      webAuthnError(1);
+    }
+  });
+}
+
+function checkError(resp) {
+  if (!('errorCode' in resp)) {
+    return false;
+  }
+  if (resp.errorCode === 0) {
+    return false;
+  }
+  webAuthnError(resp.errorCode);
+  return true;
+}
+
+function webAuthnError(errorType) {
+  const errors = {
+    browser: $('#unsupported-browser'),
+    1: $('#u2f-error-1'),
+    2: $('#u2f-error-2'),
+    3: $('#u2f-error-3'),
+    4: $('#u2f-error-4'),
+    5: $('.u2f_error_5')
```

Comment:
Two questions:

1. we still use the name `u2f`?
2. `id` or `CSS class name`? I prefer using CSS class name.

Reply:
And, in `webauthn_error.tmpl`, the names mismatch 😂